go-controller/pkg/ovn/controller/services/load_balancer.go

```diff
@@ -0,0 +1,468 @@
+							Targets: targets,
+						}
+
+						// in other words, is this ExternalTrafficPolicy=local?
+						if config.externalTrafficLocal {
+							noSNATRouterRules = append(noSNATRouterRules, rule)
+						} else {
+							routerRules = append(routerRules, rule)
+						}
+					}
+				}
+			}
+
+			// If switch and router rules are identical, coalesce
+			if reflect.DeepEqual(switchRules, routerRules) && len(switchRules) > 0 && node.gatewayRouterName != "" {
+				out = append(out, ovnlb.LB{
+					Name:        makeLBName(service, proto, "node_router+switch_"+node.name),
+					Protocol:    string(proto),
+					ExternalIDs: eids,
+					Opts:        lbOpts(service),
+					Routers:     []string{node.gatewayRouterName},
+					Switches:    []string{node.switchName},
+					Rules:       routerRules,
+				})
+			} else {
+				if len(routerRules) > 0 && node.gatewayRouterName != "" {
+					out = append(out, ovnlb.LB{
+						Name:        makeLBName(service, proto, "node_router_"+node.name),
+						Protocol:    string(proto),
+						ExternalIDs: eids,
+						Opts:        lbOpts(service),
+						Routers:     []string{node.gatewayRouterName},
+						Rules:       routerRules,
+					})
+				}
+				if len(noSNATRouterRules) > 0 && node.gatewayRouterName != "" {
+					lb := ovnlb.LB{
+						Name:        makeLBName(service, proto, "node_local_router_"+node.name),
+						Protocol:    string(proto),
+						ExternalIDs: eids,
+						Opts:        lbOpts(service),
+						Routers:     []string{node.gatewayRouterName},
+						Rules:       noSNATRouterRules,
+					}
+					lb.Opts.SkipSNAT = true
+					out = append(out, lb)
+				}
+
+				if len(switchRules) > 0 {
+					out = append(out, ovnlb.LB{
+						Name:        makeLBName(service, proto, "node_switch_"+node.name),
+						Protocol:    string(proto),
+						ExternalIDs: eids,
+						Opts:        lbOpts(service),
+						Switches:    []string{node.switchName},
+						Rules:       switchRules,
+					})
+				}
+			}
+		}
+	}
+
+	merged := mergeLBs(out)
+	if len(merged) != len(out) {
+		klog.V(5).Infof("Service %s/%s merged %d LBs to %d",
+			service.Namespace, service.Name,
+			len(out), len(merged))
+	}
+
+	return merged
+}
+
+// configsByProto buckets a list of configs by protocol (tcp, udp, sctp)
+func configsByProto(configs []lbConfig) map[v1.Protocol][]lbConfig {
+	out := map[v1.Protocol][]lbConfig{}
+	for _, config := range configs {
+		out[config.protocol] = append(out[config.protocol], config)
+	}
+	return out
+}
+
+// lbOpts generates the OVN load balancer options from the kubernetes Service.
+func lbOpts(service *v1.Service) ovnlb.LBOpts {
+	return ovnlb.LBOpts{
+		Unidling: svcNeedsIdling(service.GetAnnotations()),
+		Affinity: service.Spec.SessionAffinity == v1.ServiceAffinityClientIP,
+	}
+}
+
+// mergeLBs joins two LBs together if it is safe to do so.
+//
+// an LB can be merged if the protocol, rules, and options are the same,
+// and only the targets (switches and routers) are different.
```

Comment:
nit: "targets" is overloaded here... VIP targets(endpoints), LB targets(Where the LB is applied)

Reply:
very good point. "target" is meaningless here, deleted.